pkg/controller/secretannotator/azure/reconciler.go

```diff
@@ -150,7 +150,11 @@ func (r *ReconcileCloudCredSecret) updateSecretAnnotations(secret *corev1.Secret
 		secretAnnotations = map[string]string{}
 	}
 
-	secretAnnotations[constants.AnnotationKey] = value
+	// ARO relies on being CCO being in passthrough mode -
+	// should not automatically switch to mint mode
+	if secretAnnotations[constants.AnnotationKey] != constants.PassthroughAnnotation {
+		secretAnnotations[constants.AnnotationKey] = value
+	}
```

Comment:
I guess if a non-ARO OCP user wants to switch from passthrough to mint, they could get their creds in place and then delete the annotation but this change does make it a little less magical. What do you think @joelddiaz

Reply:
Alternative would be some kind of a config option that just says "assume passthrough". We did just recently add a configmap for another feature that could be used for this purpose.

Reply:
If to follow code behavior - No annotation meant always use Mint mode. 
This change just to preserves passthrough credentials mode. So this is "assume passthrough" configuration? isn't?

At the moment we are missing credentials checking to define if we can mint credentials. In ARO case this would not be possible so the flow would set to passthrough. So this is stopgap until this feature lands into CCO.

Reply:
> Alternative would be some kind of a config option that just says "assume passthrough". We did just recently add a configmap for another feature that could be used for this purpose.

This would cause more confusion. As we get a situation where "my annotations says Mint but behaves like Passthrough"

Reply:
If I'm reading these changes correctly, for ARO something outside of the cluster will come along an annotate the secret to force it into passthrough mode.

@mjudeikis why would it not be possible to do credentials checking in ARO? Are there Azure API calls available to allow us to perform the whole passthrough vs mint determination dynamically? Or is this a limitation of the kind of introspection the Azure API allows?

@dgoodwin I think I'd prefer the config option that just lets someone set something like 'forceMode: passthrough'. especially if we already have some external code setting the annotation. I think it makes is more clear if the change is driven through our configmap.

@mjudeikis I don't think Devan is saying having the annotation mismatch the runtime behavior. The annotation would be driven/overridden by the config option.

Reply:
@joelddiaz ARO uses the flavor of UPI, and the secret will be created with passthrough annotation in the first place. The plan is to make this option available in the Installer codebase but not used for IPI (https://github.com/jim-minter/installer/commit/588ea7de41723bcdb0ba1b30db00f79dd9a20388). 

By default, the existing code path will not change, and secrets will be created without annotation. Because ARO uses the installer as a library code we will set this during install and we need it to be persistent during install and cluster lifecycle. ConfigMap option would mean day2 operation as it is not available as an asset in the installer.

I think the configMap option is not an option here.

Happy to go via full technical details offline if this helps.

Reply:
I can see now that we force mint mode for Azure, previously I think we assumed passthrough? Did this change in 4.2 or 4.3? I assume that is where your breakage originates?

Looking at this again this morning but I still do not feel this is the right approach. We're calling a function to updateSecretAnnotations(secret *corev1.Secret, value string) and then ignoring the value if the current value happens to be passthrough.

That annotation was a programatic implementation detail, it was not meant to be a mode for an external actor to configure how the operator functions. That is a much more logical fit for some kind of configuration, not an annotation on a secret. If we implement the concept of it being something that must be preserved, it's lost it's value and we can't do what we intended with it, which was to indicate if the operator determined you can mint or not. That will be a problem when we get to the TODO on line 136 and now we can't update our annotation.

I'll be on the call today, we may be able to work out a way to get the configuration in at install time via the installer manifests directory support.

Reply:
FYI, we had identified the 'mint assumption' when the mint code originally landed https://github.com/openshift/cloud-credential-operator/pull/95#discussion_r311717042 . I can't find the approval, but I remember being told it was okay and that we'd come back and find a way to determine mint vs passthrough dynamically.

Reply:
I've worked out how to check whether minting can be enabled dynamically - see https://gist.github.com/jim-minter/de0b581fb74435eb5a602bf5b02668ae.  I think that by using that code, you can decide whether or not to set to mint mode, and that will solve the issues both for ARO and for OCP on Azure.

Reply:
(tl;dr: it turns out that all you need to do is get the oauth token, parse the JWT and inspect the roles claim)